### PR TITLE
ENH: Projections of a field's standard deviation

### PR DESCRIPTION
**Pull request by jzuhone**
## PR Summary

Weighted projections are essentially averages along the line of sight of a field. For some fields (like velocity), projecting the weighted standard deviation is also a useful quantity. This PR adds that functionality for both on-axis and off-axis projections to yt, including `ProjectionPlot`s and projections to FITS images. The functionality is demonstrated in this notebook: 

https://gist.github.com/388e2b53d1b848fc19a125669d422cec

Still need to add docs ~~and tests~~. When viewing the diff, you may want to turn off changes in whitespace since I fixed the indentation of a bunch of docstrings. 

I'm not entirely crazy about what I had to do in order to take care of cancellations in the computation of the standard deviation in the final map--basically if we end up with a negative difference between the projection of the quantity squared and the square of the projected quantity, and it can be reasonably attributed to roundoff error, we simply set the difference to zero. There are ways to compute the standard deviation which avoids the cancellation, but doing it here would actually be pretty invasive as far as the projection machinery goes. 

I also fixed a number of issues I found with documentation regarding projections. 

## PR Checklist

- [x] New features are documented, with docstrings and narrative docs
- [x] Adds a test for any bugs fixed. Adds tests for new features.

### CODE REVIEW

File: yt/data_objects/construction_data_containers.py
Comment (neutrinoceros) [inline yt/data_objects/construction_data_containers.py:538]:
```suggestion
        *,
        moment=1,
```

File: yt/data_objects/construction_data_containers.py
Comment (neutrinoceros) [inline yt/data_objects/construction_data_containers.py:438]:
```suggestion
        *,
        moment=1,
```

File: yt/visualization/plot_window.py
Comment (neutrinoceros) [inline yt/visualization/plot_window.py:2128]: is this broken ?
I personally don't fancy obscuring information from git blame with this kind of changes.

**Review by neutrinoceros — COMMENTED** [review]
super quick comments. I'm game to review this more in depth when it's ready, this looks really cool !
also FYI, CI is currently unstable (see discussion in #3988), but it should be resolved upstream in a few hours/days.

File: yt/data_objects/construction_data_containers.py
Comment (jzuhone) [reply yt/data_objects/construction_data_containers.py:538]: We need a quick explanation somewhere of why the use of the asterisk is so important and the rationale for putting kwargs before and after it is. Most developers are not fully familiar with it yet (I am myself not) and we need clarification on whether this is a style choice or a choice that the project needs to be implementing across the board.

Comment (jzuhone) [reply yt/data_objects/construction_data_containers.py:538]: I realize this is not exactly a "new" python feature either--it's just not one that we've really required yt developers to use.

Comment (neutrinoceros) [reply yt/data_objects/construction_data_containers.py:538]: It may not be new, but it *does* require Python 3, so it wasn't something we *could* use in yt until early 2020 when we dropped Python 2, which is why it's currently very sporadically used.
Anyway I'd happily open a discussion or explain my general observations/advice about this, where would be a good place for this ?

File: yt/visualization/plot_window.py
Comment (neutrinoceros) [reply yt/visualization/plot_window.py:2128]: ok then I suggest adding these commits to .git-blame-ignore-revs

File: yt/data_objects/construction_data_containers.py
Comment (jzuhone) [reply yt/data_objects/construction_data_containers.py:538]: I'd just open a github issue

Comment (neutrinoceros) [reply yt/data_objects/construction_data_containers.py:538]: here you go #3992

File: yt/funcs.py
Comment (neutrinoceros) [inline yt/funcs.py:1332]:
```suggestion
            "Cannot compute the second moment of a projection if weight_field=None!"
```

File: yt/visualization/fits_image.py
Comment (neutrinoceros) [inline yt/visualization/fits_image.py:1385]:
```suggestion
        *,
        moment=1,
```

File: yt/visualization/tests/test_offaxisprojection.py
Comment (neutrinoceros) [inline yt/visualization/tests/test_offaxisprojection.py:132]: maybe we could use a smaller buffer than the default (800x800) to mimize exec time ?

Comment (jzuhone) [reply yt/visualization/tests/test_offaxisprojection.py]: Done

File: yt/data_objects/construction_data_containers.py
Comment (neutrinoceros) [inline yt/data_objects/construction_data_containers.py:267]: This is fragile, using `functools.partial` for avoid late-binding issues, see https://github.com/yt-project/yt/pull/3996

File: yt/data_objects/construction_data_containers.py
Comment (neutrinoceros) [inline yt/data_objects/construction_data_containers.py:355]: I'm worried that this isn't necessarily sufficient to completely restore the initial state (in particular in the context of #3948). Unless there's a strong argument to do so, I'd suggest just leaving the field be.

File: yt/data_objects/data_containers.py
Comment (neutrinoceros) [inline yt/data_objects/data_containers.py:1260]:
```suggestion
    def integrate(self, field, weight=None, axis=None, *, moment=1):
```

File: yt/data_objects/construction_data_containers.py
Comment (jzuhone) [reply yt/data_objects/construction_data_containers.py:355]: We do this elsewhere, see here for the temporary field creation:

https://github.com/yt-project/yt/blob/89073a9b6d91381aff424d977491ef132427456d/yt/visualization/volume_rendering/off_axis_projection.py#L316-L337
and here for its removal:

https://github.com/yt-project/yt/blob/89073a9b6d91381aff424d977491ef132427456d/yt/visualization/volume_rendering/off_axis_projection.py#L402-L403

I'd much prefer a better way of removing fields, but that needs a separate PR. I frankly think it is better to remove it--the `FieldInfoContainer` is pretty jam-packed as is in most situations, and reducing clutter would be ideal.

Comment (neutrinoceros) [reply yt/data_objects/construction_data_containers.py:355]: I agree it shouldn't be necessary for this PR, and I wasn't aware we already had examples of this, so thank you for pointing it out. Can you open an issue to remind ourselves that we should better define/encapsulate a method to remove fields ?

File: yt/visualization/plot_window.py
Comment (jzuhone) [reply yt/visualization/plot_window.py:2137]: These changes are unfortunately interwined with other docstring edits that are part of the PR's changes itself in the same commits, so probably best to leave them alone.

Comment (neutrinoceros) [reply yt/visualization/plot_window.py:2137]: unfortunate. I'd offer to rewrite this branch' history but it doesn't seem worth the risk.

File: yt/data_objects/construction_data_containers.py
Comment (neutrinoceros) [reply yt/data_objects/construction_data_containers.py:358]: done here: #4082

**Review by neutrinoceros — DISMISSED** [review]
I guess we'd need someone else to sign off before we can merge

**Comment by jzuhone** [comment]
this needs docs--I just need time to write them up

**Comment by neutrinoceros** [comment]
oh sorry I forgot about that. I'll review again when it's done

**Comment by jzuhone** [comment]
This PR is ready for review.

File: doc/source/analyzing/objects.rst
Comment (neutrinoceros) [inline doc/source/analyzing/objects.rst:594]: should we deprecate `"mip"` ?

File: doc/source/analyzing/objects.rst
Comment (neutrinoceros) [inline doc/source/analyzing/objects.rst:666]: this paragraph is so similar to the previous one, I think we should merge them

File: doc/source/visualizing/plots.rst
Comment (neutrinoceros) [inline doc/source/visualizing/plots.rst:351]: Not sure why but the `:math:` directive didn't render correctly in the build

Screenshot 2022-09-05 at 09 20 27

File: yt/funcs.py
Comment (neutrinoceros) [inline yt/funcs.py:1337]: not sure `yt.funcs` is a good home for this function. Do you expect it to be adopted downstream ? if not, I'd suggest putting it in a more private space

File: yt/utilities/math_utils.py
Comment (neutrinoceros) [inline yt/utilities/math_utils.py:1558]: thanks a lot for this explanation, that's really helpful

File: yt/visualization/fits_image.py
Comment (neutrinoceros) [inline yt/visualization/fits_image.py:1073]: I suggest we avoid mixing two different meaning of "optional" in the context of Python function signatures.
In terms of typing, `Optional` means "possibly `None`", but here it is not the case. (FTR this rather confusing name can be avoided in Python 3.9, but for the time being we _do_ use it in yt)
```suggestion
    moment : integer
```

File: yt/visualization/fits_image.py
Comment (neutrinoceros) [inline yt/visualization/fits_image.py:1368]: same as above
```suggestion
    moment : integer
```

File: yt/visualization/fits_image.py
Comment (neutrinoceros) [inline yt/visualization/fits_image.py:1089]: For api consistency
```suggestion
        *,
        moment=1,
```

File: yt/visualization/fits_image.py
Comment (neutrinoceros) [inline yt/visualization/fits_image.py:1400]: seems like a good idea, but is there a reason for doing this change specifically in this PR that I am not seeing ?

File: yt/visualization/fits_image.py
Comment (neutrinoceros) [inline yt/visualization/fits_image.py:1405]: I don't get this change. Doesn't `_determine_fields` yield field tuples ?

File: yt/visualization/fixed_resolution.py
Comment (neutrinoceros) [inline yt/visualization/fixed_resolution.py:622]: I'm confused; I thought in most places, `fname` is meant to designate a single `str`, not the complete field tuple

File: yt/visualization/plot_window.py
Comment (neutrinoceros) [inline yt/visualization/plot_window.py:1210]:
```suggestion
                    colorbar_label = "%s \\rm{Standard Deviation}" % colorbar_label
```

File: yt/visualization/plot_window.py
Comment (neutrinoceros) [inline yt/visualization/plot_window.py:2309]:
```suggestion
        *,
        moment=1,
```

File: yt/visualization/plot_window.py
Comment (neutrinoceros) [inline yt/visualization/plot_window.py:2067]:
```suggestion
        *,
        moment=1,
```

File: yt/visualization/volume_rendering/off_axis_projection.py
Comment (neutrinoceros) [inline yt/visualization/volume_rendering/off_axis_projection.py:331]: why this change ?

**Review by neutrinoceros — COMMENTED** [review]
In general I really like this a lot. However, on a general level, I think the docs should acknowledge that this functionality is only supported for cartesian geometries at the moment. I would also recommend adding a test to check that the error message one gets when calling `moment=2` for e.g. a cylindrical geometry is sound and clear.

**Review by cphyc — COMMENTED** [review]
I haven't reviewed all files in detail but it seems to be working nicely :)
However, when saving the plot, I noticed the filename is
`{ds_name}_Projection_{axis}_{field}_{weight_field}.{extension}`.
Shouln't it be something like 
`{ds_name}_Projection_{axis}_{field}_{weight_field}_standard_deviation.{extension}`
?

File: doc/source/analyzing/objects.rst
Comment (jzuhone) [reply doc/source/analyzing/objects.rst:594]: I saw deprecation messages for "mip" already in certain places, not sure if there isn't one here already or not.

Comment (neutrinoceros) [reply doc/source/analyzing/objects.rst:594]: Oh that's right, I forgot about it. Nothing to do here !

File: doc/source/visualizing/plots.rst
Comment (jzuhone) [reply doc/source/visualizing/plots.rst:351]: Funny, this resolved OK for me on my laptop.

File: yt/visualization/fits_image.py
Comment (jzuhone) [reply yt/visualization/fits_image.py:1073]: FWIW this is part of the `numpydoc` docstring standard: 

https://numpydoc.readthedocs.io/en/latest/format.html

and we've used it in plenty of other places in yt. I wasn't aware that we started discouraging this at any point.

File: yt/visualization/fits_image.py
Comment (jzuhone) [reply yt/visualization/fits_image.py:1400]: I can't remember why I did this anymore exactly--I somewhat recall an error of some sort

File: yt/visualization/volume_rendering/off_axis_projection.py
Comment (jzuhone) [reply yt/visualization/volume_rendering/off_axis_projection.py:331]: For some reason `pre-commit` asked for it

File: yt/funcs.py
Comment (jzuhone) [reply yt/funcs.py:1337]: Do we have a standard for what should go in here? I'm honestly not certain of where else it should go. I'd obviously not want to create a single file for it.

File: yt/visualization/fixed_resolution.py
Comment (jzuhone) [reply yt/visualization/fixed_resolution.py:622]: Yes, but the problem here is that there's a name collision--`field` is already in use in the function signature. If you've got any other suggestions I'm all ears

File: doc/source/analyzing/objects.rst
Comment (jzuhone) [reply doc/source/analyzing/objects.rst:666]: Ehh. I agree that they are very similar (in fact I copied, pasted, and edited to get this paragraph). But I feel like there are competing interests here. The first and second moments are of course different, and `std` was not documented separately prior to this (an oversight I assume), so I think we should leave it as-is.

File: yt/visualization/fits_image.py
Comment (neutrinoceros) [reply yt/visualization/fits_image.py:1073]: then carry on. I'm more familiar with `typing` than I am with numpy docstring style

File: yt/data_objects/construction_data_containers.py
Comment (neutrinoceros) [inline yt/data_objects/construction_data_containers.py:186]: Sorry I missed that one
```suggestion
        *,
        moment=1,
```

File: yt/data_objects/construction_data_containers.py
Comment (neutrinoceros) [inline yt/data_objects/construction_data_containers.py:458]:
```suggestion
            moment=moment,
```

File: yt/data_objects/construction_data_containers.py
Comment (neutrinoceros) [inline yt/data_objects/construction_data_containers.py:558]:
```suggestion
            moment=moment,
```

File: yt/funcs.py
Comment (neutrinoceros) [reply yt/funcs.py:1337]: never mind, I remembered that we had a couple other `validate_*` methods but turns out they also live in this module. I've never been satisfied with how this module hosts both internal and user-facing functions but I guess it's a bit late to change it and it's certainly out of scope here

File: yt/visualization/fixed_resolution.py
Comment (neutrinoceros) [reply yt/visualization/fixed_resolution.py:622]: I think `item` would make more sense in this context; Also, just curious, I don't think I've ever seen an example of a derived field where the first argument `field` was used, do you happen to know when it's useful ?

File: yt/visualization/volume_rendering/off_axis_projection.py
Comment (neutrinoceros) [reply yt/visualization/volume_rendering/off_axis_projection.py:331]: I mean, the formatting is just `black` being triggered by the line getting long, but you're also adding a `units=""` argument, and that's what confuses me.

File: doc/source/visualizing/plots.rst
Comment (neutrinoceros) [reply doc/source/visualizing/plots.rst:351]: probably nbd then, maybe the html artefacts are just broken

File: doc/source/analyzing/objects.rst
Comment (neutrinoceros) [reply doc/source/analyzing/objects.rst:666]: ok seems fair

File: yt/visualization/volume_rendering/off_axis_projection.py
Comment (jzuhone) [reply yt/visualization/volume_rendering/off_axis_projection.py:331]: The `units` here is dimensionless, so I'm just making it explicit here.

File: yt/visualization/fits_image.py
Comment (jzuhone) [reply yt/visualization/fits_image.py:1401]: ah, now I remember--I needed this to be a data object so I could call `_determine_fields`, which we weren't doing before.

**Comment by neutrinoceros** [comment]
@jzuhone The failure is real

**Review by neutrinoceros — APPROVED** [review]

**Comment by neutrinoceros** [comment]
@cphyc do you wish to review this more in depth latter or should we advertise this some more to get a review from someone else ?

**Comment by cphyc** [comment]
I'm not sure I can review this in the coming two weeks - so if you can find an extra pair of eyes that'd be great :)

**Comment by jzuhone** [comment]
@matthewturk ?

**Review by matthewturk — APPROVED** [review]
I've reviewed these changes and they're great.  I don't see any issues that weren't addressed elsewhere.  Thank you, @jzuhone .

**Pull request merged**
